Strata/DL/Lambda/TypeFactory.lean

```diff
@@ -0,0 +1,271 @@
+A datatype description. `typeArgs` contains the free type variables of the given datatype.
+-/
+structure LDatatype (IDMeta : Type) where
+  name : String
+  typeArgs : List TyIdentifier
+  constrs: List (@LConstr IDMeta)
+deriving Repr, DecidableEq
+
+/--
+A datatype applied to arguments
+-/
+def data (d: LDatatype IDMeta) (args: List LMonoTy) : LMonoTy :=
+  .tcons d.name args
+
+/--
+The default type application for a datatype. E.g. for datatype `type List α = | Nil | Cons α (List α)`, produces LMonoTy `List α`.
+-/
+def dataDefault (d: LDatatype IDMeta) : LMonoTy :=
+  data d (d.typeArgs.map .ftvar)
+
+---------------------------------------------------------------------
+
+-- Typechecking
+
+/--
+Determines whether type name `n` appear in type `t`
+-/
+def tyNameAppearsIn (n: String) (t: LMonoTy) : Bool :=
+  match t with
+  | .tcons n1 args => n == n1 || List.any (List.map (tyNameAppearsIn n) args) id
+  | _ => false
+
+/--
+Determines whether all occurences of type name `n` within type `t` have arguments `args`. The string `c` appears only for error message information.
+-/
+def checkUniform (c: String) (n: String) (args: List LMonoTy) (t: LMonoTy) : Except Format Unit :=
+  match t with
+  | .tcons n1 args1 => if n == n1 && args == args1 then .ok ()
+    else if n == n1 then .error f!"Error in constructor {c}: Non-uniform occurrence of {n}, which is applied to {args1} when it should be applied to {args}"
+    else List.foldrM (fun t u => do
+      let _ ← checkUniform c n args t
+      .ok u
+      ) () args1
+  | _ => .ok ()
+
+
+/--
+Check for strict positivity and uniformity of datatype `d` in type `ty`. The string `c` appears only for error message information.
+-/
+def checkStrictPosUnifTy (c: String) (d: LDatatype IDMeta) (ty: LMonoTy) : Except Format Unit :=
+  match ty with
+  | .arrow t1 t2 =>
+    if tyNameAppearsIn d.name t1 then
+      .error f!"Error in constructor {c}: Non-strictly positive occurrence of {d.name} in type {ty}"
+    else checkStrictPosUnifTy c d t2
+  | _ => checkUniform c d.name (d.typeArgs.map .ftvar) ty
+
+/--
+Check for strict positivity and uniformity of a datatype
+-/
+def checkStrictPosUnif (d: LDatatype IDMeta) : Except Format Unit :=
+  List.foldrM (fun ⟨name, args⟩ _ =>
+    List.foldrM (fun ⟨ _, ty ⟩ _ =>
+      checkStrictPosUnifTy name.name d ty
+    ) () args
+  ) () d.constrs
+
+---------------------------------------------------------------------
+
+-- Generating constructors and eliminators
+
+/--
+The `LFunc` corresponding to constructor `c` of datatype `d`. Constructor functions do not have bodies or `concreteEval` functions, as they are values when applied to value arguments.
+-/
+def constrFunc (c: LConstr IDMeta) (d: LDatatype IDMeta) : LFunc IDMeta :=
+  { name := c.name, typeArgs := d.typeArgs, inputs := c.args, output := dataDefault d, isConstr := true }
+
+/--
+Generate `n` strings for argument names for the eliminator. Since there is no body, these strings do not need to be used.
```

Comment:
Probably obvious, but just noting: given that they don't need to be used, and the fact that they bind new names, it's not a problem that they might shadow existing names.